build-test.sh

```diff
@@ -684,9 +699,9 @@ fi
 # This is where all built CoreClr libraries will copied to.
 export __CMakeBinDir="$__BinDir"
 
-if [ ! -d "$__BinDir" ] || [ ! -d "$__BinDir/bin" ]; then
+if ![[ -d "$__BinDir" ] || [ -d "$__BinDir/bin" ]]; then
```

Comment:
It seems that the original logic was correct. We were failing if either one of the dirs didn't exist. In your way, we succeed even if the __BinDir/bin doesn't exist because the first conditional makes the "or" true and so the whole condition becomes false.
And I haven't seen problems building the tests recently, I've actually used this script a week or two ago just fine.

Reply:
> It seems that the original logic was correct.

It seems that it is not the case. Below you will find output of executing ./build-test.sh direct;y after repo build:

```
Repo successfully built.
Product binaries are available at /mnt/e/src/ms/dotnet/coreclr-nix/bin/Product/Linux.x64.Checked
developer@Notek-JBW10:/mnt/e/src/ms/dotnet/coreclr-nix$ ./build-test.sh x64 checked
Setting __HostDistroRid to linux-x64
Cannot find build directory for the CoreCLR Product.
Please make sure CoreCLR is built before building tests.
Example use: './build.sh x64 Checked'

// After modifying output to show  $__BinDir and __BinDir/bin values 

developer@Notek-JBW10:/mnt/e/src/ms/dotnet/coreclr-nix$ ./build-test.sh x64 checked
Setting __HostDistroRid to linux-x64
Cannot find build directory for the CoreCLR Product.
__BinDir: /mnt/e/src/ms/dotnet/coreclr-nix/bin/Product/Linux.x64.Checked
__BinDir/bin: /mnt/e/src/ms/dotnet/coreclr-nix/bin/Product/Linux.x64.Checked/bin
Please make sure CoreCLR is built before building tests.
Example use: './build.sh x64 Checked'
```

One of the checks will be always false as we do not create `bin/Product/Linux.x64.Checked/bin` directory and this will always result in logical sentence value of `true` what in turn will abort test build. The correct check should be even limited to:

```bash
if ![ -d "$__BinDir" ]; then
```

Reply:
@janvorli 

Ping. Have you had a chance to look at my answer?. I am hitting this problem on all Ubuntu VM machines I use: 16.04, 17.10, 18.04 and on Windows 10 WSL Ubuntu 16.04.

Reply:
I am sorry for the delay, I've missed your post. You are right the with just testing the _BinDir, since nothing creates _BinDir/dir ever. I wonder how come the tests build worked for me before.

Could you please update your change so that it has just that one check? And please move the ! inside the [ ] to keep the convention we use elsewhere.

Reply:
Hmm, actually, I was wrong. The _BinDir/dir actually exists, I was looking at the bin/obj/xxxxx instead of bin/Product/xxxx.
So I wonder how come this folder doesn't get created in your case. This folder contains native binaries used by the tests, like "hfa_nested_f32_native_cpp.so" etc.
And the tests need this folder, so your change is really incorrect.

Reply:
Thanks for explanation. I am checking what happens with directory structure during Linux builds on my machines

Reply:
@janvorli 

Finally I have found source of all problems. Build workflow on Linux differs from Windows in the order in which test components are built.
1. Running `./build.sh` on Linux with `-skiptests` parameter skips build of native tests (they go into `__BinDir/bin`) which are later not built when running `./build-test.sh` script
2. This is the result of moving Linux native tests build into `./build.sh` script what is breaking change for the build workflow and is not reflected in script documentation.
3. IMO we should build native test components on Linux using `./build-test.sh` command
4. As this may interfere with CI workflow where managed tests are build on Windows it would be necessary to have `-skipmanaged` switch in `./build-test.sh` to build only native test components
5. `if [ ! -d "$__BinDir" ] || [ ! -d "$__BinDir/bin" ]; then` should trigger native test components build and follow Windows workflow.

Reply:
@janvorli Ping. Could you have a look at my findings?

Reply:
@4creators thank you for looking into it. Unifying the workflow would be nice to have. Can you please create an issue to track it?

Reply:
Created #17149 to track this issue. Will update PR to reflect our discussion - remove `./build-test.sh` changes and leave Clang changes.